scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/TaskAction.java

```diff
@@ -1,163 +1,168 @@
+        private final TaskState newState;
+        
+        public ReadChangeTaskAction(WorkerConfiguration workerConfiguration, Task task, Reader reader, int tryAttempt, TaskState newState) {
+            this.workerConfiguration = Preconditions.checkNotNull(workerConfiguration);
+            this.task = Preconditions.checkNotNull(task);
+            this.reader = Preconditions.checkNotNull(reader);
+            Preconditions.checkArgument(tryAttempt >= 0);
+            this.tryAttempt = tryAttempt;
+            this.newState = newState;
+        }
+        
+        @Override
+        public CompletableFuture<TaskAction> run() {
+            if (newState != null) {
+                workerConfiguration.transport.setState(task.id, newState);
+            }
+            
+            CompletableFuture<TaskAction> taskActionFuture = reader.nextChange().
+                    thenApply(change -> new ConsumeChangeTaskAction(workerConfiguration, task, reader, change, tryAttempt));
+            return FutureUtils.thenComposeExceptionally(taskActionFuture, ex -> {
+                // Exception occured while reading the window, we will have to restart
+                // ReadNewWindowTaskAction - read a window from state defined in task.
+                long backoffTime = workerConfiguration.workerRetryBackoff.getRetryBackoffTimeMs(tryAttempt);
+                logger.atSevere().withCause(ex).log("Error while reading a CDC change. Task: %s. " +
+                        "Task state: %s. Will retry after backoff (%d ms).", task.id, task.state, backoffTime);
+                return workerConfiguration.delayedFutureService.delayedFuture(backoffTime)
+                        .thenApply(t -> new ReadNewWindowTaskAction(workerConfiguration, task, tryAttempt + 1));
+            });
+        }
+    }
+
+    private static final class ConsumeChangeTaskAction extends TaskAction {
+        private final WorkerConfiguration workerConfiguration;
+        private final Task task;
+        private final Reader reader;
+        private final Optional<RawChange> change;
+        private final int tryAttempt;
+
+        public ConsumeChangeTaskAction(WorkerConfiguration workerConfiguration, Task task, Reader reader, Optional<RawChange> change, int tryAttempt) {
+            this.workerConfiguration = Preconditions.checkNotNull(workerConfiguration);
+            this.task = Preconditions.checkNotNull(task);
+            this.reader = Preconditions.checkNotNull(reader);
+            this.change = Preconditions.checkNotNull(change);
+            Preconditions.checkArgument(tryAttempt >= 0);
+            this.tryAttempt = tryAttempt;
+        }
+
+        @Override
+        public CompletableFuture<TaskAction> run() {
+            if (change.isPresent()) {
+                Task updatedTask = task.updateState(change.get().getId());
+                CompletableFuture<TaskAction> taskActionFuture = workerConfiguration.consumer.getConsumerDispatch().consume(task, change.get())

```

Comment:
Why should Consumers decide whether `newState` is updated? (earlier in the development, Consumers were able to return their updated `TaskState`, but I decided against that - I don't think it should be their responsiblity) Wouldn't a simpler approach be to have an `if (change.get().isEndOfBatch())` here?

Reply:
End consumer does not. This is the internal consumer interface, that is wrapped in the `Consumer` class. If you utilize end-of-batch above, you are forcing state to only update on full events. Then we can't have row-by-row state management. 
Not that I am convinced it is needed, but I don't want to change things fully that much. 
This way, the wrapper created for an event consumer can ensure we get per-event state transition, and the normal ones can do it by row. 
End user consumer interfaces still return void.